hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/CallerContext.java

```diff
@@ -109,11 +114,53 @@ public String toString() {
 
   /** The caller context builder. */
   public static final class Builder {
-    private final String context;
+    private static final String KEY_VALUE_SEPARATOR = ":";
+    /**
+     * The illegal separators include '\t', '\n', '='.
+     * User should not set illegal separator.
+     */
+    private static final List<String> ILLEGAL_SEPARATORS =
+        Arrays.asList("\t","\n","=");
+    private final String fieldSeparator;
+    private final StringBuilder sb = new StringBuilder();
     private byte[] signature;
 
     public Builder(String context) {
-      this.context = context;
+      this(context, new Configuration());
+    }
+
+    public Builder(String context, Configuration conf) {
+      if (isValid(context)) {
+        sb.append(context);
+      }
+      fieldSeparator = conf.get(HADOOP_CALLER_CONTEXT_SEPARATOR_KEY,
+          HADOOP_CALLER_CONTEXT_SEPARATOR_DEFAULT);
+      checkFieldSeparator(fieldSeparator);
+    }
+
+    /**
+     * Check whether the separator is legal.
+     * The illegal separators include '\t', '\n', '='.
+     * Throw IllegalArgumentException if the separator is Illegal.
+     * @param separator the separator of fields.
+     */
+    private void checkFieldSeparator(String separator) {
+      if (ILLEGAL_SEPARATORS.stream()
```

Comment:
Not that is wrong, but we could just do: ILLEGAL_SEPARATORS.contains(separator), and if we make it a HashSet would be faster.

Reply:
Ok. It's done